apps/fft/src/fft.fil

```diff
@@ -0,0 +1,187 @@
+
+      let start = reg_i*2*N;
+      let end = (reg_i+1)*2*N;
+
+      // sync in -> unsync in
+      c1 := new CombineButterflyBundle[N, W]<'G + sync_in_t>(sync_data_in{r_stage}{start..end}{0..2});
+      r1 := R<'G + sync_in_t, 'G + usync_out_t>(c1.out);
+      s1 := new SplitButterflyBundle[N, W, usync_out_t-usync_in_t]<'G + usync_in_t>(r1.out);
+      // unsync in -> butterfly
+      b := B<'G + usync_in_t>(s1.out{0..2*N}{0..2}, twiddle_in{r_stage}{reg_i*N..(reg_i+1)*N}{0..2});
+      // unsync out -> sync out
+      c2 := new CombineButterflyBundle[N, W]<'G + usync_out_t>(b.out{0..2*N}{0..2});
+      r2 := R<'G + usync_out_t, 'G + sync_out_t>(c2.out);
+      s2 := new SplitButterflyBundle[N, W]<'G + sync_out_t-1>(r2.out);
+
+      // store sync out
+      sync_data_out{r_stage}{start..end}{0..2} = s2.out{0..2*N}{0..2};
+    } 
+  }
+
+  // --------------- STRIDE PERMUTATIONS ----------------
+  SP := new StridePermutation[NStages, W] in ['G+stage_latency-1, 'G+latency];
+  
+  // A delay is necessary for between stages so the pass-through register
+  // isn't used twice at the same time
+  D := new Delay[W*NPoints*2];
+
+
+  // hook up stride permutations
+  for stage in 0..NStages {
+    let sync_out_t = (stage + 1) * stage_latency - 1;
+    stride := SP<'G + sync_out_t>(sync_data_out{stage}{0..NPoints}{0..2});
+
+    // delay between stages
+    c1 := new CombineButterflyBundle[NPoints/2, W]<'G + sync_out_t>(stride.out{0..NPoints}{0..2});
+    d1 := D<'G + sync_out_t>(c1.out);
+    s1 := new SplitButterflyBundle[NPoints/2, W]<'G + sync_out_t + 1>(d1.out);
+
+    sync_data_in{stage+1}{0..NPoints}{0..2} = s1.out{0..NPoints}{0..2};
+  }
+
+  out{0..NPoints}{0..2} = sync_data_in{NStages}{0..NPoints}{0..2};
+  L := latency;
+}
+
+/// Computes N butterflies at once.
+comp Butterflies[N, W, E, ?M=W-E-1]<'G: L>(
+  go: interface['G],
+  in[N*2][2]: ['G, 'G+L] W, // Each butterfly takes 2 complex numbers
+  twiddle[N][2]: ['G, 'G+L] W // twiddle factor for each butterfly
+) -> (
+  out[N*2][2]: ['G+L, 'G+L+1] W
+) with {
+  some L where L > 0;
+} where
+    N > 0,
+    W == E + M + 1,
+    E > 1,
+    M > 1
+{
+  Bx := new Butterfly[W, E, M];
+  L := Bx::L;
+
+  for i in 0..N {
+    b := new Butterfly[W, E, M]<'G>(in{2*i}{0..2}, in{2*i+1}{0..2}, twiddle{i}{0..2});
+    out{2*i}{0..2} = b.out0{0..2};
+    out{2*i+1}{0..2} = b.out1{0..2};
+  }
+}
+
+/// Combines output of the Butterflies component into a single wire
+comp CombineButterflyBundle[N, W, ?L=1]<'G: L>(
+  in[N*2][2]: ['G, 'G+L] W
+) -> (
+  out: ['G, 'G+L] W*4*N
+) where
+    N > 0,
+    W > 0,
+    L > 0
+{
+  bundle outb[N*4]: ['G, 'G+L] W;
+  for i in 0..N*2 {
+    outb{i*2} = in{i}{0};
+    outb{i*2+1} = in{i}{1};
+  }
+  
+  cb := new ConcatBundle[W, N*4, L]<'G>(outb{0..N*4});
+  
+  out = cb.out;
+}
+
+/// Splits output of the Butterflies component into a bundle
+comp SplitButterflyBundle[N, W, ?L=1]<'G: L>(
```

Comment:
This is the opposite operation: Partitioning. Also worth being in reshape library one day!